File: packages/core/src/feeds/service.ts
import type { Session } from '../auth'
import type { Paginated } from '../common'
import type { PluginRegistry, ResponseParser, Scraper } from '../utils'
import type { HttpClient } from '../utils'
import { FeedNotFoundError } from './error'
import type { FeedsRepository } from './repository'
import type { ExtractedFeed, Feed, FeedDto, ProcessedFeed } from './types'

export class FeedsService {
	constructor(
		private readonly repo: FeedsRepository,
		private readonly http: HttpClient,
		private readonly parser: ResponseParser<Document>,
		private readonly registry: PluginRegistry<ExtractedFeed, ProcessedFeed>,
		private readonly defaultScraper: Scraper<ExtractedFeed, ProcessedFeed>,
	) {}

	async list(session: Session): Promise<Paginated<Feed>> {
		const feeds = await this.repo.findMany({
			profileId: session.profileId,
		})

		return {
			hasMore: false,
			data: feeds,
		}
	}

	async get(id: string, session: Session): Promise<Feed> {
		const feed = await this.repo.findOne({
			id,
			profileId: session.profileId,
		})
		if (!feed) {
			throw new FeedNotFoundError(id)
		}

		return feed
	}

	async create(dto: FeedDto, session: Session): Promise<Feed> {
		const scraper =
			this.registry.load(new URL(dto.url).hostname) ?? this.defaultScraper

		const req = scraper.prepare(dto.url)

		const res = await this.http.get(req)
		const document = await this.parser.parse(res)

		const extracted = scraper.extract(dto.url, document)
		const processed = scraper.postprocess(dto.url, extracted)

		return this.repo.create({
			feedUrl: dto.url,
			processed,
			profileId: session.profileId,
		})
	}

	async delete(id: string, session: Session): Promise<void> {
		const deleted = await this.repo.delete({
			id,
			profileId: session.profileId,
		})
		if (!deleted) {
			throw new FeedNotFoundError(id)
		}
	}
}

File: crates/core/src/feeds/service.rs
use std::sync::Arc;

use super::{model::CreateFeed, Error, FeedFindManyParams, FeedsRepository, ProcessedFeed};
use crate::{
    common::{FindOneParams, Paginated, Session},
    feeds::FeedCreateData,
    scraper::Scraper,
    Feed,
};

pub struct FeedsService {
    feeds_repo: Arc<dyn FeedsRepository + Send + Sync>,
    scraper: Arc<dyn Scraper<ProcessedFeed> + Send + Sync>,
}

impl FeedsService {
    pub fn new(
        feeds_repo: Arc<dyn FeedsRepository + Send + Sync>,
        scraper: Arc<dyn Scraper<ProcessedFeed> + Send + Sync>,
    ) -> Self {
        Self {
            feeds_repo,
            scraper,
        }
    }

    pub async fn list(&self, session: Session) -> Result<Paginated<Feed>, Error> {
        let params = FeedFindManyParams {
            profile_id: session.profile_id,
        };
        let feeds = self.feeds_repo.find_many(params).await?;

        let paginated = Paginated::<Feed> {
            has_more: false,
            data: feeds,
        };

        Ok(paginated)
    }

    pub async fn get(&self, id: String, session: Session) -> Result<Feed, Error> {
        let params = FindOneParams {
            id,
            profile_id: session.profile_id,
        };
        let feed = self.feeds_repo.find_one(params).await?;

        Ok(feed)
    }

    pub async fn create(&self, data: CreateFeed, session: Session) -> Result<Feed, Error> {
        let scraped = self.scraper.scrape(&data.url).await?;

        let data = FeedCreateData {
            url: data.url,
            feed: scraped,
            profile_id: session.profile_id,
        };
        let feed = self.feeds_repo.create(data).await?;

        Ok(feed)
    }

    pub async fn delete(&self, id: String, session: Session) -> Result<(), Error> {
        let params = FindOneParams {
            id,
            profile_id: session.profile_id,
        };
        self.feeds_repo.delete(params).await?;

        Ok(())
    }
}
